### ISSUE TITLE
Possible Ionic 4 integrations? Updated

### ISSUE THREAD

**Issue by ariffromeo** [kamiazya/ngx-face-api-js#14]
Hey @kamiazya  it is wonderful to have this working on Angular. Is it possible to integrate this module to Ionic 4? I tried to implement all the steps in Ionic 4 as how it is implemented in your Angular sample [here](https://www.npmjs.com/package/ngx-face-api-js), but it doesnt work. Compiled without any errors, but the detection canvas not appear. can u help me with this issue?

**My update on 17 September '19 - (Found a workaround, but not fully)**

After a while i tried to find the reason why the attributes 'allFaces' and others arent working in Ionic. It seems that those attributes wouldn't bind together inside `<ion-router-outlet></ion-router-outlet>`. Inside `<ion-router-outlet>` is where all the Ionic components such as headers and content will be placed (inside main.page.html). 

![index-page-html](https://user-images.githubusercontent.com/8829689/65010793-8fcb1c80-d943-11e9-80ba-147daa4aa438.JPG)

The basic ionic structure is like this `<ion-app><ion-router-outlet></ion-router-outlet></ion-app>`. Then i tried to write the same code outside `<ion-router-outlet>` which is inside `<ion-app></ion-app>` tag and bam its working! 

![index-page-html2](https://user-images.githubusercontent.com/8829689/65011178-28ae6780-d945-11e9-9290-41599b88620f.JPG)

Plus I noticed that in your Angular samples, the attributes binding work inside the `<my-app></my-app>` tag which is almost similar like Ionic structure (we know Ionic was built on top of Angular). So can you make a fix or any workaround to make this working inside Ionic? Thank you

**Comment by CodingApp** [comment]
**@ariffromeo**
Thank you for sharing this, while installing and following the same guide i'm getting these errors. Please take a look. Thank you :)
![Capture2](https://user-images.githubusercontent.com/44866144/65119404-a5087f80-da05-11e9-9a38-26df466d8278.PNG)
![Capture](https://user-images.githubusercontent.com/44866144/65119423-a9cd3380-da05-11e9-874f-f3597c962529.PNG)

**Comment by ariffromeo** [comment]
> **@ariffromeo**
> Thank you for sharing this, while installing and following the same guide i'm getting these errors. Please take a look. Thank you :)
> ![Capture2](https://user-images.githubusercontent.com/44866144/65119404-a5087f80-da05-11e9-9a38-26df466d8278.PNG)
> ![Capture](https://user-images.githubusercontent.com/44866144/65119423-a9cd3380-da05-11e9-874f-f3597c962529.PNG)

hey @CodingApp make sure you do `import { NgxFaceApiJsModule } from 'ngx-face-api-js';` inside `app.module.ts` and import `NgxFaceApiJsModule.forRoot({ modelsUrl: 'https://raw.githubusercontent.com/justadudewhohacks/face-api.js/master/weights',})` inside `@NgModule`
![index-page-html3](https://user-images.githubusercontent.com/8829689/65134811-7efbd300-da37-11e9-990a-7e969d1d6966.JPG)

if the problem still occur, make sure your Angular version is 7.

**Comment by CodingApp** [comment]
I have import well it in app component like you discuss above. But still i'm getting this weird errors in terminal. Can you please look in to it. Thank you.

![Capture](https://user-images.githubusercontent.com/44866144/65139577-53242180-da26-11e9-93f6-22fb3ed8fac8.PNG)

**Comment by ariffromeo** [comment]
> I have import well it in app component like you discuss above. But still i'm getting this weird errors in terminal. Can you please look in to it. Thank you.
> 
> ![Capture](https://user-images.githubusercontent.com/44866144/65139577-53242180-da26-11e9-93f6-22fb3ed8fac8.PNG)

It's like u install tfjs into ur project. in my case, install tfjs also wouldnt work (will probably try in future) but for now i didn't. it will also work without tfs (instead knowing that having them will improve the speed). what i did to make it work, downgrade my angular to 7 n follow the guides and using my way above.

**Comment by kamiazya** [comment]
> > I have import well it in app component like you discuss above. But still i'm getting this weird errors in terminal. Can you please look in to it. Thank you.
> > ![Capture](https://user-images.githubusercontent.com/44866144/65139577-53242180-da26-11e9-93f6-22fb3ed8fac8.PNG)
> 
> It's like u installed tfjs into ur project. in my case, installing tfjs also wouldnt work (will probably try in future) but for now i didn't. it will also work without tfs (instead knowing that having them will improve the speed). what i did to make it work, downgrade my angular to 7 n follow the guides and using my way above.

I also encountered an error in Angular 8 when I created the verification environment.

I looked into the fece-api.js repository for similar errors and found that it seems to be an error caused by the TypeScript version.

https://github.com/justadudewhohacks/face-api.js/issues/415

As @ariffromeo  says, you need to downgrade to Angular 7.

**Comment by kamiazya** [comment]
Since it is difficult to build a verification environment for ionic 4 and Angular 7, can someone create a verification repository and invite me?

Perhaps it may be easier to support Angular 8, so in parallel, I would like to proceed with support so that it works with Angular 8 in another issue.

**Comment by ariffromeo** [comment]
> Since it is difficult to build a verification environment for ionic 4 and Angular 7, can someone create a verification repository and invite me?
> 
> Perhaps it may be easier to support Angular 8, so in parallel, I would like to proceed with support so that it works with Angular 8 in another issue.

I can do a repo for my working project using this directives. 

oh @kamiazya i have a **another question** too, In my project `home.page.ts`, i have codes that can generate a` <video allFaces></video>` tag element onto HTML through a div with an id `<div #videoContainer></div>` after compiled.

In my HTML, if i do straight using `<video allFaces></video>` it will work and give face detection output. **However** if use `<div #videoContainer></div>` (later will replaced by after compiled) it will not work. how can i solve this? Hopefully u can help me as soon as u are free :)

**my html**
![index-page-html4](https://user-images.githubusercontent.com/8829689/65209507-f169c300-daca-11e9-812e-3c61466868a8.JPG)

**after compiled**
![index-page-html5](https://user-images.githubusercontent.com/8829689/65209842-d21f6580-dacb-11e9-973e-2d83b7e3c417.JPG)

**Comment by CodingApp** [comment]
> > I have import well it in app component like you discuss above. But still i'm getting this weird errors in terminal. Can you please look in to it. Thank you.
> > ![Capture](https://user-images.githubusercontent.com/44866144/65139577-53242180-da26-11e9-93f6-22fb3ed8fac8.PNG)
> 
> It's like u installed tfjs into ur project. in my case, installing tfjs also wouldnt work (will probably try in future) but for now i didn't. it will also work without tfs (instead knowing that having them will improve the speed). what i did to make it work, downgrade my angular to 7 n follow the guides and using my way above.

These are the setups i have followed to get all of these errors:
1. First setup blank Ionic 4 project. (Angular version is 7.x.x not 8)
2. Secondly installed this package using npm install -S @angular/cdk ngx-face-api-js face-api.js (I have installed all the packages that are related with ngx-face-api.js)
3. I have commented code in add code in app component html and also include it like discuss above.
4. **I'm getting error while i open mode links in direct like some thing error 404**

Some body please discuss that do i create a sample repo using angular 7 and let you to check it for me and might help somebody in future :) 

Thanks

**Comment by kamiazya** [comment]
> > > I have import well it in app component like you discuss above. But still i'm getting this weird errors in terminal. Can you please look in to it. Thank you.
> > > ![Capture](https://user-images.githubusercontent.com/44866144/65139577-53242180-da26-11e9-93f6-22fb3ed8fac8.PNG)
> > 
> > 
> > It's like u installed tfjs into ur project. in my case, installing tfjs also wouldnt work (will probably try in future) but for now i didn't. it will also work without tfs (instead knowing that having them will improve the speed). what i did to make it work, downgrade my angular to 7 n follow the guides and using my way above.
> 
> These are the setups i have followed to get all of these errors:
> 
> 1. First setup blank Ionic 4 project. (Angular version is 7.x.x not 8)
> 2. Secondly installed this package using npm install -S @angular/cdk ngx-face-api-js face-api.js (I have installed all the packages that are related with ngx-face-api.js)
> 3. I have commented code in add code in app component html and also include it like discuss above.
> 4. **I'm getting error while i open mode links in direct like some thing error 404**
> 
> Some body please discuss that do i create a sample repo using angular 7 and let you to check it for me and might help somebody in future :)
> 
> Thanks

@CodingApp 
Thank you for the detailed procedure.

Can you tell me the version of ionic cli that can set up Step 1 ionic 4 and Angular 7 projects?
I will try it too.

Also, since the TypeScript version seems to be affecting the error, can you try some of the TypeScript versions?

Thanks

**Comment by CodingApp** [comment]
**kamiazya**

![Capture](https://user-images.githubusercontent.com/44866144/65216606-f6c70d80-daca-11e9-8387-7fe5347e0fef.PNG)

I'm using many ionic 3 and ionic 4 projects using the same setting along with react native too don't know how to handle this one. Its good if somebody look into it. Thank you

**Comment by ariffromeo** [comment]
I have created a repo for my working Ionic files here  [https://github.com/ariffromeo/Ionic4-ngx-face-api-js](url)

**Comment by CodingApp** [comment]
**@ariffromeo**

Thank you so much hopefully i would also help many of user in future too :)

**Comment by ariffromeo** [comment]
@CodingApp @kamiazya please try clone and do 'npm install' for all the modules first before ionic serve..

**Comment by ariffromeo** [comment]
> > Since it is difficult to build a verification environment for ionic 4 and Angular 7, can someone create a verification repository and invite me?
> > Perhaps it may be easier to support Angular 8, so in parallel, I would like to proceed with support so that it works with Angular 8 in another issue.
> 
> I can do a repo for my working project using this directives.
> 
> oh @kamiazya i have a **another question** too, In my project `home.page.ts`, i have codes that can generate a` <video allFaces></video>` tag element onto HTML through a div with an id `<div #videoContainer></div>` after compiled.
> 
> In my HTML, if i do straight using `<video allFaces></video>` it will work and give face detection output. **However** if use `<div #videoContainer></div>` (later will replaced by after compiled) it will not work. how can i solve this? Hopefully u can help me as soon as u are free :)
> 
> **my html**
> ![index-page-html4](https://user-images.githubusercontent.com/8829689/65209507-f169c300-daca-11e9-812e-3c61466868a8.JPG)
> 
> **after compiled**
> ![index-page-html5](https://user-images.githubusercontent.com/8829689/65209842-d21f6580-dacb-11e9-973e-2d83b7e3c417.JPG)

@kamiazya can u help me with this? :)

**Comment by kamiazya** [comment]
> > > Since it is difficult to build a verification environment for ionic 4 and Angular 7, can someone create a verification repository and invite me?
> > > Perhaps it may be easier to support Angular 8, so in parallel, I would like to proceed with support so that it works with Angular 8 in another issue.
> > 
> > 
> > I can do a repo for my working project using this directives.
> > oh @kamiazya i have a **another question** too, In my project `home.page.ts`, i have codes that can generate a` <video allFaces></video>` tag element onto HTML through a div with an id `<div #videoContainer></div>` after compiled.
> > In my HTML, if i do straight using `<video allFaces></video>` it will work and give face detection output. **However** if use `<div #videoContainer></div>` (later will replaced by after compiled) it will not work. how can i solve this? Hopefully u can help me as soon as u are free :)
> > **my html**
> > ![index-page-html4](https://user-images.githubusercontent.com/8829689/65209507-f169c300-daca-11e9-812e-3c61466868a8.JPG)
> > **after compiled**
> > ![index-page-html5](https://user-images.githubusercontent.com/8829689/65209842-d21f6580-dacb-11e9-973e-2d83b7e3c417.JPG)
> 
> @kamiazya can u help me with this? :)

@ariffromeo

Sorry, I'm not good at English, so I don't understand the situation.
(This sentence is also generated by Google Translate.)

If this question is completely different from the main issue of this issue, it would be helpful to have the issue set up in the verification repository you created earlier.
It would be helpful if there was a branch that could reproduce the same situation.

**Comment by kamiazya** [comment]
I tried it in the repository for verification and found that it was necessary to import NgxFaceApiModule separately in AppModule child modules (HomeModule etc.).

However, if you import it in a child module, another error will occur, so it cannot be done with the current version.

From now on, try to stop dividing the module and avoid this problem.

**Comment by kamiazya** [comment]
I did well!

I issued a Pull Request, so please check this solution.

https://github.com/ariffromeo/Ionic4-ngx-face-api-js/pull/2

<img width="622" alt="Ionic_App" src="https://user-images.githubusercontent.com/35218186/65252612-3ece4a80-db34-11e9-8024-daea45fc1421.png">

And, ionic 4 also (but limitedly) proved that this module can be used, so this issue will be closed.

Some issues found in this issue will be addressed in another issue.


If you have any questions about this matter, we will continue to address this issue.

Thanks

**Comment by ariffromeo** [comment]
Thank you for the solution @kamiazya now it works wihtin ionic 4 and angular 8. 

I understand only `<img>` and `<video>` tags in HTML are working to bind the attribute. 

**My Problem**  : Now i have `<video>` tag is generated **dynamically** using the code `this.video = document.createElement('video');` and i attached `allFaces` in here `this.video.setAttributeNS(null, 'allFaces', '');`. 

When i compiled it will produce `<video allFaces autoplay height="480" width="640"></video>`, but the detection output is **not working**.the **CDK canvas** also **not appear**..please @kamiazya help me on this. because it is important for my project :)

-------------------- ** @kamiazya  i tried my best to translate this into Japanese**----------------------

HTMLの<img>および<video>タグのみが属性バインディングに機能することを理解しています。

私の問題：<video>タグは、this.video = document.createElement（ 'video'）;というコードを使用して動的に生成されます。 そして、this.video.setAttributeNS（null、 'allFaces'、 ''）をallFacesに添付しました。

コンパイルすると、<video allFaces autoplay height = "480" width = "640"> </ video>が生成されますが、検出出力は機能しません。 CDKキャンバスも表示されません。 それは私のプロジェクトにとって重要だからです:)

**Comment by kamiazya** [comment]
> Thank you for the solution @kamiazya now it works wihtin ionic 4 and angular 8.
> 
> I understand only <img> and <video> tags in HTML are working to bind the attribute.
> 
> My Problem : Now i have <video> tag is generated dynamically using the code this.video = document.createElement('video'); and i attached allFaces in here this.video.setAttributeNS(null, 'allFaces', '');.
> 
> When i compiled it will produce <video allFaces autoplay height="480" width="640"></video>, but the detection output is not working.the CDK canvas also not appear..please @kamiazya help me on this. because it is important for my project :)
> 
> -------------------- ** @kamiazya i tried my best to translate this into Japanese**----------------------
> 
> HTMLのおよびタグのみが属性バインディングに機能することを理解しています。
> 
> 私の問題：タグは、this.video = document.createElement（ 'video'）;というコードを使用して動的に生成されます。 そして、this.video.setAttributeNS（null、 'allFaces'、 ''）をallFacesに添付しました。
> 
> コンパイルすると、 </ video>が生成されますが、検出出力は機能しません。 CDKキャンバスも表示されません。 それは私のプロジェクトにとって重要だからです:)

thanks for translation.
I understand the situation.

Basically, if it is not a video tag generated by Angular, you can not receive the function provided by the module even if you add directives.

If it is important to embed video tags dynamically, you may be able to do it using Angular's Dynamic Component Loader pattern.

https://angular.io/guide/dynamic-component-loader
